Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,158 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A store for instrumentation App Ids. This makes sure we don't query the public endpoint to find an app Id for the same Ikey more than once.
+    /// </summary>
+    internal static class CorrelationIdLookupHelper
```

Comment:
It's static because I wanted to share the dictionary (mapping between instrumentation key and app ID) across everything. 
Besides, the dictionary, there aren't any instance specific properties. 

One can argue whether or not we should make the endpoint an instance specific property. Perhaps - we can and avoid  some computation - but then again the current code is optimized such that it will only worry about the endpoint if the instrumentation key is not found in the current dictionary entries - so not too worried about that either.